csrc/kernel_cache.cpp

```diff
@@ -472,255 +452,165 @@ void FusionKernelRuntime::prepareRuntimeOrder() {
 }
 
 // passing args by value, since we will be modify this
-void FusionKernelRuntime::startAsyncCompile(KernelArgumentHolder& args_old) {
+void FusionKernelRuntime::startAsyncCompile(
+    const KernelArgumentHolder& input_args) {
   // only single compilation is supported at this moment.
   std::unique_lock<std::mutex> unique_lock(mutex_, std::try_to_lock);
   TORCH_CHECK(
       unique_lock.owns_lock(),
-      "Calling startAsyncCompile on a FusionKernelRuntime that's already starting a compilation thread is not supported");
+      "Calling startAsyncCompile on a FusionKernelRuntime that has already",
+      " started a compilation thread is not supported.",
+      " - unique_lock");
   std::unique_lock<std::mutex> unique_lock2(compiling_, std::try_to_lock);
   TORCH_CHECK(
       unique_lock2.owns_lock(),
-      "Calling startAsyncCompile on a FusionKernelRuntime that's already starting a compilation thread is not supported 2");
-
-  // for some reason I can't seem to move unique_lock and it keeps using copy.
-  // auto compile_fusion = [args = std::move(args_old), lock =
-  // std::move(unique_lock), this] () mutable {
-  auto compile_fusion = [args = std::move(args_old), this]() mutable {
+      "Calling startAsyncCompile on a FusionKernelRuntime that has already",
+      " started a compilation thread is not supported.",
+      " - unique_lock2");
+
+  // TODO: Compilation can happen in parallel! We can infer the output sizes
+  // on the non-compiled kernel and build the entire tensor_map prior to
+  // asyc compilation.
+
+  // PyTorch's threadpool uses std::function, which requires the target to be
+  // copy-constructible. Adding a std::unique_lock to the lambda's capture list
+  // prevents it from being copyable. The std::unique_lock can be moved, but it
+  // cannot be copied. Thus, we need the second mutex.
+  auto compile_fusion = [args = input_args, this]() mutable {
```

Comment:
Why `mutable`?

Reply:
Since `args` is modified in the lambda and is captured, you need the `mutable` keyword.
The lambda calls `runSegmentsWithInputs(KernelArgumentHolder& args, ...)`

Reply:
Hmm, I didn't know `mutable` is necessary in this case. Thanks.